### ISSUE TITLE
Keras training problem - no overfit with a small number of data

### ISSUE THREAD

**Issue by mortezamg63** [keras-team/keras#14606]
I am new to Keras. I have a problem in training the defined model. It is hard to find the issue in my code. My implementation for a model on recommendation system data does not work correctly because there is no overfit. In my code, I wrote a custom data generator that convert data to one-hot encoding and hashing encoding. Then it uses one-hot and hashed features for training the model. The data generator gets features and converts the sample data to one-hot encoded vector. Then, it is used to declare train and validation data loaders. Please look at the code and short description in the following. Let me know what is wrong in my code.
```
class my_custom_dataset_generator(tf.keras.utils.Sequence):
def __init__(self,data, Mu_NumericalFs, Sigma_NumericalFs, Uniq_Features, batch_size,hash_bins):
    self.rawdata = data
...
def __getitem__(self,index):
// process/generate one-hot and hash encoding for self.rawdata[index]. Then return the vector
```

The defined model has two hidden layers (Dense/fully connected layers) and because it is a binary classification problem I have one or two neurons with binary cross entropy or categorical cross entropy loss functions in the output layer.

For training, I used 200 samples out of 400 (here for showing the unexpected result). I train model and it must overfit but there is no overfit. The train and validation accuracies are following the same trend. It is confusing. In the following, you can see what I have done for training:

# Training phase

```
train = dataset[:200]
valid = dataset[200:400]

 train_ds =my_custom_dataset_generator(train,mu,sigma,feats,batchsize,bin_hash_size)
 valid_ds = my_custom_dataset_generator(valid,mu,sigma,feats,batchsize,bin_hash_size)
 EC = model(input_shape,512,256,256,2,0.1)
 #data here are processed  to get mean, variance and number of categories (feats) in each column of 
 dataset.

for epoch in range(start, 15):
    data=[]
    label = []
    train_acc = []
    train_loss = []
    pbar = tq.tqdm(enumerate(train_ds), total=len(train_ds))
    for i,d in pbar:
        data.append(d[0])
        label.append(d[1])
        if (i+1)%batchsize==0:  *#batchsize=16*
            history = EC.fit(np.array(data),np.array(label),epochs=1,verbose=0)#,batch_size=16)
            data = []
            label = []
            train_acc.append(history.history['accuracy'])
            train_loss.append(history.history['loss'])
     train_acc = np.array(train_acc)
     train_loss = np.array(train_loss)
     print("[class-Train] - Accuracy:", train_acc.mean()," , Loss: ", train_loss.mean())

```

I did the same for validation and got data from valid_ds.

As you can see, in the data generator I get one sample of data and append it to the list. Then list is used as a batch to train the model.

I changed data generator and I get a batch of data from data generator. It is standard way of doing the job. You can check how I implemented the new data generator based on tutorial keras-generate-data. In the new implementation I just call model.fit() without for loop. I have overfit in this version of my implementation.

I really appreciate let me know why here I do not have overfit in the code above. But in the new of implementation I have overfit (the code below).

Here is the code (overfit occurrence):

```
class my_custom_dataset_generator(tf.keras.utils.Sequence):
   def __init__(self,data, Mu_NumericalFs, Sigma_NumericalFs, Uniq_Features, batch_size,hash_bins,outputLabel=False):
self.rawdata = data


   def __getitem__(self,index):
       train, label = [],[]
       for idx in range(index*self.batch_size, (index+1)*self.batch_size):
           # process/generate one-hot and hash encoding for self.rawdata[idx]. Then return the vector
           ....
           train.append(one-hot vector)
           label.append(target)
       return train, label




def model(n_dim, n_hidden1, n_hidden2, n_hidden3, n_classes,rec_w):
   input_layer = tf.keras.layers.Input(shape=(n_dim,))
   encoded = tf.keras.layers.Dense(n_hidden1, activation='relu',name='E1',kernel_regularizer=tf.keras.regularizers.l2(1e-4))(input_layer)
   encoded = tf.keras.layers.Dropout(0.4)(encoded)
   encoded = tf.keras.layers.Dense(n_hidden2, activation='relu',name='E2',kernel_regularizer=tf.keras.regularizers.l2(1e-4))(encoded)

   classifier = tf.keras.layers.Dense(n_classes, activation='softmax',name='Classification')(encoded)


   logloss = tf.keras.losses.BinaryCrossentropy(from_logits=False)
   ssautoencoder = tf.keras.models.Model(inputs=[input_layer], outputs=[classifier])
   ssautoencoder.compile(optimizer='adam', loss=[logloss],metrics=['accuracy'])
   return ssautoencoder 

train = dataset[:200]
valid = dataset[200:400]

 train_ds =my_custom_dataset_generator(train,mu,sigma,feats,batchsize,bin_hash_size)
 valid_ds = my_custom_dataset_generator(valid,mu,sigma,feats,batchsize,bin_hash_size)
 EC = model(input_shape,512,256,256,2,0.1)

EC.fit_generator(generator=train_ds, validation_data=valid_ds,epochs=10)
```
